xbmc/games/controllers/windows/GUIPortList.cpp

```diff
@@ -0,0 +1,319 @@
+#include "games/controllers/types/ControllerHub.h"
+#include "games/controllers/types/ControllerTree.h"
+#include "games/controllers/types/PortNode.h"
+#include "guilib/GUIMessage.h"
+#include "guilib/GUIWindow.h"
+#include "guilib/LocalizeStrings.h"
+#include "messaging/ApplicationMessenger.h"
+#include "messaging/helpers/DialogOKHelper.h"
+#include "utils/StringUtils.h"
+#include "utils/log.h"
+#include "view/GUIViewControl.h"
+#include "view/ViewState.h"
+
+using namespace KODI;
+using namespace ADDON;
+using namespace GAME;
+
+CGUIPortList::CGUIPortList(CGUIWindow& window)
+  : m_guiWindow(window),
+    m_viewControl(std::make_unique<CGUIViewControl>()),
+    m_vecItems(std::make_unique<CFileItemList>())
+{
+}
+
+CGUIPortList::~CGUIPortList()
+{
+  Deinitialize();
+}
+
+void CGUIPortList::OnWindowLoaded()
+{
+  m_viewControl->Reset();
+  m_viewControl->SetParentWindow(m_guiWindow.GetID());
+  m_viewControl->AddView(m_guiWindow.GetControl(CONTROL_PORT_LIST));
+}
+
+void CGUIPortList::OnWindowUnload()
+{
+  m_viewControl->Reset();
+}
+
+bool CGUIPortList::Initialize(GameClientPtr gameClient)
+{
+  m_gameClient = std::move(gameClient);
+
+  m_viewControl->SetCurrentView(DEFAULT_VIEW_LIST);
+
+  Refresh();
+
+  CServiceBroker::GetAddonMgr().Events().Subscribe(this, &CGUIPortList::OnEvent);
+
+  return true;
+}
+
+void CGUIPortList::Deinitialize()
+{
+  CServiceBroker::GetAddonMgr().Events().Unsubscribe(this);
+
+  CleanupItems();
+
+  m_gameClient.reset();
+}
+
+bool CGUIPortList::HasControl(int controlId)
+{
+  return m_viewControl->HasControl(controlId);
+}
+
+int CGUIPortList::GetCurrentControl()
+{
+  return m_viewControl->GetCurrentControl();
+}
+
+void CGUIPortList::Refresh()
+{
+  // Send a message to clear the view control
+  m_viewControl->Clear();
+
+  CleanupItems();
+
+  unsigned int itemIndex = 0;
+  for (const CPortNode& port : m_gameClient->Input().GetActiveControllerTree().GetPorts())
+    AddItems(port, itemIndex, GetLabel(port));
+
+  m_viewControl->SetItems(*m_vecItems);
+
+  // Try to restore focus to the previously focused port
+  if (!m_focusedPort.empty() && m_addressToItem.find(m_focusedPort) != m_addressToItem.end())
+  {
+    const unsigned int itemIndex = m_addressToItem[m_focusedPort];
+    m_viewControl->SetSelectedItem(itemIndex);
+    OnItemFocus(itemIndex);
+  }
+}
+
+void CGUIPortList::FrameMove()
+{
+  const int itemIndex = m_viewControl->GetSelectedItem();
+  if (itemIndex != m_currentItem)
+  {
+    m_currentItem = itemIndex;
+    if (itemIndex >= 0)
+      OnItemFocus(static_cast<unsigned int>(itemIndex));
+  }
+}
+
+void CGUIPortList::SetFocused()
+{
+  m_viewControl->SetFocused();
+}
+
+void CGUIPortList::OnSelect()
+{
+  const int itemIndex = m_viewControl->GetSelectedItem();
+  if (itemIndex >= 0)
+    OnItemSelect(static_cast<unsigned int>(itemIndex));
+}
+
+void CGUIPortList::ResetPorts()
+{
+  m_gameClient->Input().ResetPorts();
+
+  // Refresh the GUI
+  using namespace MESSAGING;
```

Comment:
Interesting placement, I guess it's scoped to the method?

Reply:
Scoped to the stanza, substantively.